### DISCUSSION TITLE
Is this board supported?

### DISCUSSION THREAD

**Discussion by g1r1** [meerk40t/meerk40t#1709] [Q&A] [answered]
Hello all. Was looking at buying a CO2 laser with this board. 

https://imgur.com/a/GWN5tdB

It does look like a Moshidraw board but I am not sure because it also has bluetooth support. 

Will this board be supported? Any clues?

**Comment by tatarize** [comment, marked as answer]
Currently, no. We know it has a bluetooth chip but we don't know the software it takes or whether the protocol is similar to the regular moshiboard. At a minimum we can tell that that the interface chip is not a CH341x so that means that at a minimum it cannot work out of the box.

We looked for a MS107 but couldn't find one. We're currently working on some NewlyDraw drivers for the EnBinV8 boards. Best case that basically runs the same as Moshi but has a serial connection or something. But, even in the best case it can't be supported out of the box. Combine that with the ease of buying an M3 Nano or GRBL card and the software that gives you and we've not really seen more than about 3 of them. We think there's some part of the world that has a lot more of them though, but we can't seem to acquire one to at least check how hard it would be to write a driver for.

**Reply by g1r1** [reply]
Thanks for the quick response.  I understand; I did search for MS107 too but couldn't find references to it; the only places I could find them were some sellers selling spares like this https://ocamachinedelhi.com/product/pcb-of-co2-machine/ (this is without bluetooth but the same MS107). The other clue I was being told is that it runs software, with screenshots below; not sure that will help though.

https://imgur.com/a/dzRr0GF

Meanwhile, I will try to hunt down a machine with a board that is compatible, which isn't easy because most of the sellers have no clue what they running inside.

**Reply by joerlane** [reply]
Do you have an M107 board already? 

I would be interested to make it work; if you do have one on hand... 

Where are you located? Those M107 boards seem particularly rare in the US.

**Reply by g1r1** [reply]
No, I don't have the board. I was considering support before buying; I might still take a plunge if the seller offers a good price, though. FYI, I am in Asia.

**Comment by tatarize** [comment]
![usLBsSu](https://github.com/meerk40t/meerk40t/assets/3302478/f7754b7f-359f-43a4-8321-1c684360b6b5)

**Reply by adden00** [reply]
Hi! Have you tried to connect with this board? Was it successful? I got the same laser with this board, and any attempts to connect it with USB wasn't success. PC recognize it only like USB-storage,  and don't work. 

It only works with chinese mobile app. I tried to analyze it, but didn`t understand. Or maybe you know what commands language it works? I tried to send raw HPGL and G-code commands, and it didn't work..(  I tried to send by bluetooth directly, and by usb from phone..

**Reply by tatarize** [reply]
It connects with blue tooth. I might be able to figure out the language since I'm sort of an expert at that. If you have some caps of the mobile device communicating with the board I could probably narrow it down quite significantly.

But, no, I never managed to purchase one of these boards and consequently couldn't do any probing.

**Reply by adden00** [reply]
I sent to tatarize@gmail.com

**Reply by tatarize** [reply]
Educated guesses.

Status is 2 bytes, eg.
2066
2032
2022
6066

Data seems to be in 16 byte chunks. The end bits are always 000102030405060708090a0b0c0d0e0f and some number of bytes get rewritten.
6c0e2b00031d00ca0010002b00050250
260e00a2475c0000853d000050000088
df0e153b095000081f00d0074006ce1f
ec0e023913aa1b021613ae1202f412d5
350e0202d912feec01bd12d7d301aa12
f20efab5019a12d594018e12ea70018c
3b0e12ae4f018e12ea2d019a12f51001
1a0eaa12eaf600bd12aee000d912bad1
c80e00f412aec7001613aac3003913fe
f10ec7005d13fad1007c13aae0009a13
6d0ebef600b313ba1001c913fe2d01d9
b00e13ba4f01e313f77001e613be9401
4f0ee313ddb501d913fdd301c913faec
1f0e01b313ee02029a13ae12027c13f7
8b0e1b025d13ff1f023913f71b021613
920ed51b02131375172d0000f5190200
320e00f5ad010a00ee49012900ffee00
df0e5b00f7c5007a00f59d009d00ba7c
160e00c300fe5d00ec00f52b004701fd
4f0e0c00ad01ba00001902be00001c13
9d016200ad01ba00001902be00001c13 different command code
3d046112b100ba00001902be00001c13 different command code.
0b0eaa0c008813fe2b00ec13f75d0049
9b0e14ea7c007214ee9d009814bac500
4f0ebb14ffee00db14ea49010c15d7ad
1b0e012b15fa19023515be172d3515be
7a0e842d2b15d5e82d0c15be442edb14
3d0ed76e2ebb14df932e9814ffb72e72
4b0e14fad62e4914fd072fec13dd272f
ad0e8813df312f1c13df312f1902fa27
ac0e2fad01ee072f4701ead62eec00ae
ea0eb72ec300ff932e9d00aa6e2e7a00
440ed5442e5b00eae82d2900fa842d0a
1b0e00f7172d0000aaf02c00005f0000
7f0900008080808080808000005f0000
9d0162008080808080808000005f0000
3d0361238080808080808000005f0000
9d0162238080808080808000005f0000
* 134 repeats.

Codewise most of these have a character 01 that is almost always 0E for most of the codes. This is probably checksum followed by 0E each packet. If we delete those we get another pattern appearing at 0x05. This seems to be a 2 two-byte numbers and some other value. Notice how the second value is 0x12 repeatedly and last value is nearly always 01 this is probably `<command> x, y`

Offset(h) 00 01 02 03 04

00000000  15 3B 09 50 00  .;.P.
00000005  08 1F 00 D0 07  ...Ð.
0000000A  40 06 CE 1F 02  @.Î..
0000000F  39 13 AA 1B 02  9.ª..
00000014  16 13 AE 12 02  ..®..
00000019  F4 12 D5 02 02  ô.Õ..
0000001E  D9 12 FE EC 01  Ù.þì.
00000023  BD 12 D7 D3 01  ½.×Ó.
00000028  AA 12 FA B5 01  ª.úµ.
0000002D  9A 12 D5 94 01  š.Õ”.
00000032  8E 12 EA 70 01  Ž.êp.
00000037  8C 12 AE 4F 01  Œ.®O.
0000003C  8E 12 EA 2D 01  Ž.ê-.
00000041  9A 12 F5 10 01  š.õ..
00000046  AA 12 EA F6 00  ª.êö.
0000004B  BD 12 AE E0 00  ½.®à.
00000050  D9 12 BA D1 00  Ù.ºÑ.
00000055  F4 12 AE C7 00  ô.®Ç.
0000005A  16 13 AA C3 00  ..ªÃ.
0000005F  39 13 FE C7 00  9.þÇ.
00000064  5D 13 FA D1 00  ].úÑ.
00000069  7C 13 AA E0 00  |.ªà.
0000006E  9A 13 BE F6 00  š.¾ö.
00000073  B3 13 BA 10 01  ³.º..
00000078  C9 13 FE 2D 01  É.þ-.
0000007D  D9 13 BA 4F 01  Ù.ºO.
00000082  E3 13 F7 70 01  ã.÷p.
00000087  E6 13 BE 94 01  æ.¾”.
0000008C  E3 13 DD B5 01  ã.Ýµ.
00000091  D9 13 FD D3 01  Ù.ýÓ.
00000096  C9 13 FA EC 01  É.úì.
0000009B  B3 13 EE 02 02  ³.î..
000000A0  9A 13 AE 12 02  š.®..
000000A5  7C 13 F7 1B 02  |.÷..
000000AA  5D 13 FF 1F 02  ].ÿ..
000000AF  39 13 F7 1B 02  9.÷..
000000B4  16 13 D5 1B 02  ..Õ..
000000B9  13 13 75 17 2D  ..u.-
000000BE  00 00 F5 19 02  ..õ..
000000C3  00 00 F5 AD 01  ..õ..
000000C8  0A 00 EE 49 01  ..îI.
000000CD  29 00 FF EE 00  ).ÿî.
000000D2  5B 00 F7 C5 00  [.÷Å.
000000D7  7A 00 F5 9D 00  z.õ..
000000DC  9D 00 BA 7C 00  ..º|.
000000E1  C3 00 FE 5D 00  Ã.þ].
000000E6  EC 00 F5 2B 00  ì.õ+.
000000EB  47 01 FD        G.ý

![image](https://github.com/meerk40t/meerk40t/assets/3302478/afed3216-eedb-4850-987e-f3657cd5121f)

00 is the most common followed by 13 there.

This appears to be non-swizzled command code in a `cmd, x (2 bytes), y (2 bytes). Sort of format.

**Reply by tatarize** [reply]
The packets are checksum, length, and 15 characters or so.

6c0e2b00031d00ca0010002b00050250

6C is checksum, though it doesn't match to anything known. And 0E... 15 characters of payload.

Some of the weirder ones are single character commands.

3d0361238080808080808000005f0000
So this is 3d and then 03 ... which has a command of `61 23 80`

**Reply by tatarize** [reply]
The command codes in the middle where the command is almost certainly go here and cut this are:
0xAA 3
0xAE 6
0xBA 3
0xBE 2
0xD5 2
0xD7
0xDD
0xEA 3
0xEE
0xF5
0xF7
0xFA 3
0xFD
0xFE 3

That's a fairly random assessment of 14 different values. Which is actually pretty much what Moshi original had. We're assuming that's CUT_XY. We are missing 0xDF and 0xFF but this is actually the swizzle value 15. Which is actually MOSHI_CUT_ABS = 15.

Yeah, that's Moshi-code.

**Reply by tatarize** [reply]
Let me try decoding this.
---

9d0162030405060708090a0b0c0d0e0f

This is 9d checksum. The command length is 01 which issues the command 62. Command 62: which is in group 12. Group 12 was previously unused.

---

610261ff0005060708090a0b0c0d0e0f

This is 61 checksum. The command length is 02 which issues the command 61ff. Command 61 is group 4. In programmed mode that would do RasterSpeed but as a realtime command was previously unused.

---

7b036123c605060708090a0b0c0d0e0f

This is 7b checksum. The command length is 03 which issues the command 6123c6. Again Command 61 group 4, previously unused.

---

b80461123400060708090a0b0c0d0e0f

This is b8 checksum. The command length is 04 which issues the command 61123400 This is again a command 61.

---

6c0e2b00031d00ca0010002b00050250

This is checksum 6c command length is 0e which is the rest of the packet: 2b00031d00ca0010002b00050250. 2b is command group 0.

This does SET_OFFSET. And takes 3 16 values. z=0003, x=1d00, y=ca00. Command 10 is group 0 again. z = 002b, x=0005, y=0250

---

260e00a2475c0000853d000050000088

This is checksum 26 command length 0e which is the rest of the packet: 00a2475c0000853d000050000088. 00 is group 0, with the offsets of: z = a247, x = 5c00, y = 0085. Command 3d is group 3. Which does Moshi Move Vert.

I'm guessing somewhere this stopped working. And the code is a bit or a lot different.

---

420e1527065000081f00d00740065f00
d80a0000002727272727272740065f00

This is another packet of data. 1527065000081f00d00740065f00. Then actually just 0xa bits of data. 00000027272727272727. 

The end of that packet is just a bunch of 27 commands which is group 3, called TERMINATION. And should occur 7 times and very much does.

So 5F is MOVE_ABS, and this goes to 0000 0000. So this is an offset Goto 0 packet. Which is fairly knowable. 

---

9d016200002727272727272740065f00

This is another command 62, length 1

The previous one was:

9d0162030405060708090a0b0c0d0e0f

---

Which had 9d for the checksum. And the rest of the packet trash which is identical between commands, probably because it's reused and knows nothing will be used from that. So the checksum of 62 is 9d or the checksum of `6201`

The checksum xor of those two is exactly 0xFF. We know that the checksum of 61ff. is just 61. This tends to imply this isn't a checksum but an 0xFF-XOR hash. 

Testing this on 420e1527065000081f00d00740065f00
hex(0xFF ^ 0x15 ^ 0x27 ^ 0x06 ^ 0x50 ^ 0x00 ^ 0x08 ^ 0x1f ^ 0x00 ^ 0xd0 ^ 0x07 ^ 0x40 ^ 0x06 ^ 0x5f ^ 0x00)
'0x42'
